// @ant-design/pro-chat/es/components/BackToBottomButton/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["target", "visibilityHeight"];
import { VerticalAlignBottomOutlined } from '@ant-design/icons';
import { FloatButton } from 'antd';
import React, { useEffect, useState } from 'react';
import getScroll from "./utils/getScroll";
import throttleByAnimationFrame from "./utils/throttleByAnimationFrame";
import { jsx as _jsx } from "react/jsx-runtime";
export var BackToBottomButton = function BackToBottomButton(props) {
  var target = props.target,
    _props$visibilityHeig = props.visibilityHeight,
    visibilityHeight = _props$visibilityHeig === void 0 ? 50 : _props$visibilityHeig,
    rest = _objectWithoutProperties(props, _excluded);
  var _useState = useState(visibilityHeight === 0),
    _useState2 = _slicedToArray(_useState, 2),
    visible = _useState2[0],
    setVisible = _useState2[1];
  var handleScroll = throttleByAnimationFrame(function (e) {
    var _e$target, _e$target2;
    var scrollTop = getScroll(e.target, true);
    var scrollHeight = (_e$target = e.target) === null || _e$target === void 0 ? void 0 : _e$target.scrollHeight;
    var targetHeight = (_e$target2 = e.target) === null || _e$target2 === void 0 ? void 0 : _e$target2.clientHeight;
    setVisible(scrollHeight - scrollTop - targetHeight >= visibilityHeight);
  });
  var internalRef = React.useRef(null);
  var getDefaultTarget = function getDefaultTarget() {
    return internalRef.current && internalRef.current.ownerDocument ? internalRef.current.ownerDocument : window;
  };
  useEffect(function () {
    var getTarget = target || getDefaultTarget;
    var container = getTarget();
    handleScroll({
      target: container
    });
    container === null || container === void 0 || container.addEventListener('scroll', handleScroll);
    return function () {
      handleScroll.cancel();
      container === null || container === void 0 || container.removeEventListener('scroll', handleScroll);
    };
  }, [target]);
  return /*#__PURE__*/_jsx(FloatButton, _objectSpread({
    ref: internalRef,
    style: {
      bottom: (props.offsetTop || 148) + 20,
      position: 'absolute',
      display: visible ? 'block' : 'none'
    },
    icon: /*#__PURE__*/_jsx(VerticalAlignBottomOutlined, {})
  }, rest));
};
