// @ant-design/pro-chat/es/components/Animation/AnimText.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { animate, motion, useMotionValue, useTransform } from 'framer-motion';
import { useEffect, useState } from 'react';
import { jsx as _jsx } from "react/jsx-runtime";
export function AnimText(_ref) {
  var children = _ref.children,
    Render = _ref.Render;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    done = _useState2[0],
    setDone = _useState2[1];
  var baseText = children;
  var count = useMotionValue(1);
  var rounded = useTransform(count, function (latest) {
    return Math.round(latest);
  });
  var displayText = useTransform(rounded, function (latest) {
    return baseText.slice(0, latest);
  });
  useEffect(function () {
    setDone(false);
    var conText = animate(count, baseText.length, {
      type: 'tween',
      delay: -0.1,
      duration: 1,
      ease: 'easeInOut',
      onComplete: function onComplete() {
        setDone(true);
      }
    });
    return function () {
      try {
        conText.stop();
      } catch (e) {}
    };
  }, [baseText]);
  return done ? /*#__PURE__*/_jsx(Render, {
    children: baseText
  }) : /*#__PURE__*/_jsx(motion.div, {
    style: {
      lineHeight: 1.8
    },
    children: displayText
  });
}
