// @ant-design/pro-chat/es/components/ChatList/Messages/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { AnimText } from "../../Animation/AnimText";
import { LOADING_FLAT } from "../../../const/message";
import { Collapse, Divider, Typography } from 'antd';
import { memo } from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import BubblesLoading from "../Loading";
import { useStyles } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
var MemoHr = /*#__PURE__*/memo(function (props) {
  return /*#__PURE__*/_jsx(Divider, _objectSpread({
    style: {
      marginBottom: '1em',
      marginTop: 0
    }
  }, props));
});
var MemoDetails = /*#__PURE__*/memo(function (props) {
  return /*#__PURE__*/_jsx(Collapse, _objectSpread({
    style: {
      marginBottom: '1em'
    }
  }, props));
});
var MemoImage = /*#__PURE__*/memo(function (props) {
  return /*#__PURE__*/_jsx("img", _objectSpread({}, props));
});
var MemoAlink = /*#__PURE__*/memo(function (props) {
  return /*#__PURE__*/_jsx(Typography.Link, _objectSpread({}, props));
});
var components = {
  details: MemoDetails,
  hr: MemoHr,
  a: MemoAlink,
  img: MemoImage
  // pre: Code,
};
var Render = function Render(_ref) {
  var children = _ref.children;
  var _useStyles = useStyles(),
    styles = _useStyles.styles;
  return /*#__PURE__*/_jsx(ReactMarkdown, {
    remarkPlugins: [remarkGfm],
    className: styles.markdown,
    components: components,
    children: children
  });
};
export var MessageComponent = /*#__PURE__*/memo(function (_ref2) {
  var content = _ref2.content,
    animation = _ref2.animation;
  if (content === LOADING_FLAT) return /*#__PURE__*/_jsx(BubblesLoading, {});
  if (typeof content !== 'string') return content;
  if (!animation) {
    return /*#__PURE__*/_jsx(Render, {
      children: content
    });
  }
  return /*#__PURE__*/_jsx(Typography, {
    children: /*#__PURE__*/_jsx(AnimText, {
      Render: Render,
      children: content
    })
  });
});
