// @ant-design/pro-chat/es/utils/useStyle/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { useStyleRegister as useDefaultStyleRegister } from '@ant-design/cssinjs';
import { ConfigProvider as AntdConfigProvider, theme as antdTheme } from 'antd';
import { useContext } from 'react';
export var resetComponent = function resetComponent(token) {
  return {
    boxSizing: 'border-box',
    margin: 0,
    padding: 0,
    color: token.colorText,
    fontSize: token.fontSize,
    lineHeight: token.lineHeight,
    listStyle: 'none'
  };
};
export var operationUnit = function operationUnit(token) {
  return {
    // FIXME: This use link but is a operation unit. Seems should be a colorPrimary.
    // And Typography use this to generate link style which should not do this.
    color: token.colorLink,
    outline: 'none',
    cursor: 'pointer',
    transition: "color ".concat(token.motionDurationSlow),
    '&:focus, &:hover': {
      color: token.colorLinkHover
    },
    '&:active': {
      color: token.colorLinkActive
    }
  };
};

/**
 * 封装了一下 antd 的 useStyle，支持了一下antd@4
 * @param componentName {string} 组件的名字
 * @param styleFn {GenerateStyle} 生成样式的函数
 * @returns UseStyleResult
 */
export function useProChatStyleRegister(componentName, styleFn) {
  var _antdTheme$useToken = antdTheme.useToken(),
    token = _antdTheme$useToken.token,
    theme = _antdTheme$useToken.theme,
    hashId = _antdTheme$useToken.hashId;
  var chatToken = _objectSpread(_objectSpread({}, token), {}, {
    chatCls: '',
    antCls: ''
  });
  var _useContext = useContext(AntdConfigProvider.ConfigContext),
    getPrefixCls = _useContext.getPrefixCls;
  chatToken.chatCls = ".".concat(getPrefixCls('pro-chat'));
  chatToken.antCls = ".".concat(getPrefixCls());
  return {
    wrapSSR: useDefaultStyleRegister({
      theme: theme,
      token: token,
      hashId: hashId,
      path: ["ProChat".concat(componentName)]
    }, function () {
      return styleFn(token);
    }),
    hashId: hashId
  };
}
