// @ant-design/pro-chat/es/components/ProChatTitle.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { formatTime } from "../utils/formatTime";
import { resetComponent, useProChatStyleRegister } from "../utils/useStyle";
import { Flex } from 'antd';
import cx from 'classnames';

/**
 * Props for the ProChatTitle component.
 */
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var genStyle = function genStyle(token) {
  return _defineProperty({}, token.componentCls, {
    fontSize: 12,
    color: token.colorTextDescription,
    lineHeight: 1,
    display: 'flex',
    alignItems: 'center',
    '&-time': {
      visibility: 'hidden',
      color: token.colorTextSecondary,
      opacity: 0,
      lineHeight: 1,
      transition: 'all 0.3s ' + token.motionEaseInOut
    },
    '&:hover': _defineProperty({}, "".concat(token.componentCls, "-time"), {
      visibility: 'visible'
    })
  });
};

/**
 * ProChat 的标题组件
 * @param prefixCls
 * @returns
 */
function useStyle(prefixCls) {
  return useProChatStyleRegister('Title', function (token) {
    var proChatToken = _objectSpread(_objectSpread({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genStyle(proChatToken), resetComponent(proChatToken)];
  });
}

/**
 * prochat 的标题组件
 * @param param0
 * @returns
 */
export var ProChatTitle = function ProChatTitle(_ref2) {
  var style = _ref2.style,
    prefixClass = _ref2.prefixClass,
    className = _ref2.className,
    placement = _ref2.placement,
    time = _ref2.time,
    title = _ref2.title;
  var _useStyle = useStyle(prefixClass),
    wrapSSR = _useStyle.wrapSSR,
    hashId = _useStyle.hashId;
  return wrapSSR( /*#__PURE__*/_jsxs(Flex, {
    className: cx(hashId, prefixClass, className),
    style: _objectSpread({
      flexDirection: placement === 'left' ? 'row' : 'row-reverse'
    }, style),
    gap: 8,
    children: [title, time && /*#__PURE__*/_jsx("time", {
      className: cx("".concat(prefixClass, "-time"), hashId),
      children: formatTime(time)
    })]
  }));
};
