// @ant-design/pro-chat/es/utils/fetch.js
import _regeneratorRuntime from "@babel/runtime/helpers/esm/regeneratorRuntime";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
export var getMessageError = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(response) {
    var chatMessageError;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          chatMessageError = {
            message: "response error, status: ".concat(response.statusText),
            type: response.status
          };
          return _context.abrupt("return", chatMessageError);
        case 2:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function getMessageError(_x) {
    return _ref.apply(this, arguments);
  };
}();
var waitTime = function waitTime(time) {
  return new Promise(function (resolve) {
    setTimeout(resolve, time);
  });
};

/**
 * 使用流式方法获取数据
 * @param fetchFn
 * @param options
 */
export var processSSE = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee3(response) {
    var _options$signal;
    var options,
      _options$onErrorHandl,
      chatMessageError,
      returnRes,
      data,
      reader,
      decoder,
      finishText,
      done,
      _options$onMessageHan,
      _yield$reader$read,
      value,
      doneReading,
      chunkValue,
      _options$onFinish,
      _args3 = arguments;
    return _regeneratorRuntime().wrap(function _callee3$(_context3) {
      while (1) switch (_context3.prev = _context3.next) {
        case 0:
          options = _args3.length > 1 && _args3[1] !== undefined ? _args3[1] : {};
          if (response.ok) {
            _context3.next = 7;
            break;
          }
          _context3.next = 4;
          return getMessageError(response);
        case 4:
          chatMessageError = _context3.sent;
          (_options$onErrorHandl = options.onErrorHandle) === null || _options$onErrorHandl === void 0 || _options$onErrorHandl.call(options, chatMessageError);
          return _context3.abrupt("return");
        case 7:
          returnRes = response.clone();
          data = response.body;
          if (data) {
            _context3.next = 11;
            break;
          }
          return _context3.abrupt("return");
        case 11:
          reader = data.getReader();
          decoder = new TextDecoder();
          finishText = '';
          done = false;
          options === null || options === void 0 || (_options$signal = options.signal) === null || _options$signal === void 0 || _options$signal.addEventListener('abort', /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2() {
            return _regeneratorRuntime().wrap(function _callee2$(_context2) {
              while (1) switch (_context2.prev = _context2.next) {
                case 0:
                  reader.cancel();
                case 1:
                case "end":
                  return _context2.stop();
              }
            }, _callee2);
          })));
        case 16:
          if (done) {
            _context3.next = 30;
            break;
          }
          _context3.next = 19;
          return reader.read();
        case 19:
          _yield$reader$read = _context3.sent;
          value = _yield$reader$read.value;
          doneReading = _yield$reader$read.done;
          done = doneReading;
          chunkValue = decoder.decode(value, {
            stream: !doneReading
          });
          finishText += chunkValue;
          _context3.next = 27;
          return waitTime(0);
        case 27:
          (_options$onMessageHan = options.onMessageHandle) === null || _options$onMessageHan === void 0 || _options$onMessageHan.call(options, chunkValue, returnRes, done ? 'done' : 'progress');
          _context3.next = 16;
          break;
        case 30:
          if (done) {
            (_options$onFinish = options.onFinish) === null || _options$onFinish === void 0 || _options$onFinish.call(options, finishText, 'done');
          }
          return _context3.abrupt("return", returnRes);
        case 32:
        case "end":
          return _context3.stop();
      }
    }, _callee3);
  }));
  return function processSSE(_x2) {
    return _ref2.apply(this, arguments);
  };
}();
