// @ant-design/pro-chat/es/components/ChatItem/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { useContext } from 'react';
import { ConfigProvider, Flex } from 'antd';
import cx from 'classnames';
import AnimationItem from "../Animation";
import { ProChatAvatar } from "../ProChatAvatar";
import { ProChatTitle } from "../ProChatTitle";
import { useStyle } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
import { Fragment as _Fragment } from "react/jsx-runtime";
var runRender = function runRender(render, props, defaultDom) {
  if (render) {
    for (var _len = arguments.length, rest = new Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
      rest[_key - 3] = arguments[_key];
    }
    return render.apply(void 0, [props, defaultDom].concat(rest));
  }
  return defaultDom;
};

/**
 * Represents a chat item component.
 *
 * @component
 * @example
 * ```tsx
 * <ChatItem
 *   onAvatarClick={handleAvatarClick}
 *   className="chat-item"
 *   placement="left"
 *   avatar={avatarData}
 *   style={itemStyle}
 *   time={messageTime}
 *   messageExtra={extraContent}
 *   chatListItemContentStyle={contentStyle}
 *   chatListItemTitleStyle={titleStyle}
 *   chatItemRenderConfig={renderConfig}
 *   chatListItemAvatarStyle={avatarStyle}
 *   chatListItemExtraStyle={extraStyle}
 *   onDoubleClick={handleDoubleClick}
 * >
 *   {messageContent}
 * </ChatItem>
 * ```
 *
 * @param {ChatItemProps} props - The props for the ChatItem component.
 * @returns {React.ReactElement} The rendered ChatItem component.
 */
export var ChatItem = function ChatItem(props) {
  var _chatItemRenderConfig;
  var onAvatarClick = props.onAvatarClick,
    className = props.className,
    children = props.children,
    _props$placement = props.placement,
    placement = _props$placement === void 0 ? 'left' : _props$placement,
    avatar = props.avatar,
    style = props.style,
    time = props.time,
    animation = props.animation,
    contentAfter = props.contentAfter,
    contentBefore = props.contentBefore,
    chatListItemAvatarClassName = props.chatListItemAvatarClassName,
    chatListItemContentClassName = props.chatListItemContentClassName,
    chatListItemTitleClassName = props.chatListItemTitleClassName,
    chatListItemContentStyle = props.chatListItemContentStyle,
    chatListItemTitleStyle = props.chatListItemTitleStyle,
    chatItemRenderConfig = props.chatItemRenderConfig,
    chatListItemAvatarStyle = props.chatListItemAvatarStyle,
    chatListItemExtraStyle = props.chatListItemExtraStyle;
  var _useContext = useContext(ConfigProvider.ConfigContext),
    getPrefixCls = _useContext.getPrefixCls;
  var prefixClass = getPrefixCls('pro-chat-list-item');
  var _useStyle = useStyle(prefixClass),
    wrapSSR = _useStyle.wrapSSR,
    hashId = _useStyle.hashId;
  if ((chatItemRenderConfig === null || chatItemRenderConfig === void 0 ? void 0 : chatItemRenderConfig.render) === false) return null;
  var titleDom = runRender(chatItemRenderConfig === null || chatItemRenderConfig === void 0 ? void 0 : chatItemRenderConfig.titleRender, props, /*#__PURE__*/_jsx(ProChatTitle, {
    className: chatListItemTitleClassName,
    style: chatListItemTitleStyle,
    prefixClass: cx("".concat(prefixClass, "-message-title")),
    title: avatar === null || avatar === void 0 ? void 0 : avatar.title,
    placement: placement,
    time: time
  }));
  var avatarDom = runRender(chatItemRenderConfig === null || chatItemRenderConfig === void 0 ? void 0 : chatItemRenderConfig.avatarRender, props, /*#__PURE__*/_jsx(ProChatAvatar, {
    className: chatListItemAvatarClassName,
    avatar: avatar === null || avatar === void 0 ? void 0 : avatar.avatar,
    background: avatar === null || avatar === void 0 ? void 0 : avatar.backgroundColor,
    title: avatar === null || avatar === void 0 ? void 0 : avatar.title,
    onClick: onAvatarClick,
    prefixCls: "".concat(prefixClass, "-message-avatar"),
    style: chatListItemAvatarStyle
  }));
  var childrenDom = runRender(chatItemRenderConfig === null || chatItemRenderConfig === void 0 ? void 0 : chatItemRenderConfig.contentRender, props, children);
  var contentBeforeRender = runRender(chatItemRenderConfig === null || chatItemRenderConfig === void 0 ? void 0 : chatItemRenderConfig.contentBeforeRender, props, contentBefore);
  var contentAfterRender = runRender(chatItemRenderConfig === null || chatItemRenderConfig === void 0 ? void 0 : chatItemRenderConfig.contentAfterRender, props, contentAfter);
  var itemDom = wrapSSR( /*#__PURE__*/_jsxs(Flex, {
    className: cx(prefixClass, hashId, "".concat(prefixClass, "-").concat(placement), className),
    style: style,
    vertical: true,
    gap: 12,
    children: [titleDom, /*#__PURE__*/_jsxs(Flex, {
      style: _objectSpread({
        flexDirection: placement === 'left' ? 'row' : 'row-reverse'
      }, style),
      gap: 8,
      align: 'flex-start',
      className: cx("".concat(prefixClass, "-message-container"), hashId),
      children: [avatarDom, /*#__PURE__*/_jsx(Flex, {
        vertical: true,
        gap: 4,
        className: cx("".concat(prefixClass, "-message-container"), "".concat(prefixClass, "-message-container-").concat(placement), hashId),
        children: /*#__PURE__*/_jsxs("div", {
          children: [contentBeforeRender ? /*#__PURE__*/_jsx("div", {
            className: cx("".concat(prefixClass, "-message-before"), "".concat(prefixClass, "-message-before-").concat(placement), hashId),
            style: chatListItemExtraStyle,
            children: contentBeforeRender
          }) : null, /*#__PURE__*/_jsx(AnimationItem, {
            animation: animation,
            className: cx("".concat(prefixClass, "-message-content"), "".concat(prefixClass, "-message-content-").concat(placement), chatListItemContentClassName, hashId),
            style: chatListItemContentStyle,
            children: childrenDom
          }), contentAfterRender ? /*#__PURE__*/_jsx("div", {
            className: cx("".concat(prefixClass, "-message-after"), "".concat(prefixClass, "-message-after-").concat(placement), hashId),
            style: chatListItemExtraStyle,
            children: contentAfterRender
          }) : null]
        })
      })]
    })]
  }));
  return /*#__PURE__*/_jsx(_Fragment, {
    children: (chatItemRenderConfig === null || chatItemRenderConfig === void 0 || (_chatItemRenderConfig = chatItemRenderConfig.render) === null || _chatItemRenderConfig === void 0 ? void 0 : _chatItemRenderConfig.call(chatItemRenderConfig, props, {
      avatar: /*#__PURE__*/_jsx(ProChatAvatar, {
        avatar: avatar === null || avatar === void 0 ? void 0 : avatar.avatar,
        title: avatar === null || avatar === void 0 ? void 0 : avatar.title
      }),
      title: /*#__PURE__*/_jsx(ProChatTitle, {
        title: avatar === null || avatar === void 0 ? void 0 : avatar.title,
        time: time
      }),
      messageContent: children,
      itemDom: itemDom
    }, itemDom)) || itemDom
  });
};
