// @ant-design/pro-chat/es/components/ChatItem/style.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { resetComponent, useProChatStyleRegister } from "../../utils/useStyle";
var genStyle = function genStyle(token) {
  return _defineProperty({}, token.componentCls, {
    position: 'relative',
    maxWidth: '100vw',
    padding: '16px',
    contentVisibility: 'auto',
    cursor: 'pointer',
    containIntrinsicSize: '100px',
    '&:hover': _defineProperty({}, "".concat(token.componentCls, "-message-title-time"), {
      visibility: 'visible',
      opacity: 1
    }),
    '&-message-avatar': {
      width: 40,
      height: 40
    },
    '&-message-avatar-emoji': {
      fontSize: 24,
      textAlign: 'center',
      lineHeight: '36px'
    },
    '&-message-container': {
      width: '100%',
      display: 'flex',
      '&-right': {
        alignItems: 'flex-end'
      },
      '&-left': {
        alignItems: 'flex-start'
      }
    },
    '&-message-content': {
      width: 'max-content',
      maxWidth: '100vw',
      fontSize: token.fontSize,
      padding: token.paddingXS,
      border: "1px solid ".concat(token.colorBorderSecondary),
      borderRadius: token.borderRadius,
      backgroundColor: token.colorBgContainer,
      p: resetComponent(token),
      pre: resetComponent(token)
    }
  });
};

/**
 * ProchatItem
 * @param prefixCls
 * @returns
 */
export function useStyle(prefixCls) {
  return useProChatStyleRegister('ListItem', function (token) {
    var proChatToken = _objectSpread(_objectSpread({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genStyle(proChatToken), resetComponent(proChatToken)];
  });
}
