// @ant-design/pro-chat/es/components/ChatList/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import SkeletonList from "./SkeletonList";
import { useContext } from 'react';
import { DEFAULT_AVATAR, DEFAULT_USER_AVATAR } from "../../const/meta";
import { ConfigProvider } from 'antd';
import { cx } from 'antd-style';
import { ChatItem } from "../ChatItem";
import { MessageComponent } from "./Messages";
import { useStyle } from "./style";
import { jsx as _jsx } from "react/jsx-runtime";
var ChatList = function ChatList(props) {
  var chatListRef = props.chatListRef,
    chatItemRenderConfig = props.chatItemRenderConfig,
    className = props.className,
    loading = props.loading,
    chatListItemContentStyle = props.chatListItemContentStyle,
    chatListItemTitleStyle = props.chatListItemTitleStyle,
    chatListLeftItemContentStyle = props.chatListLeftItemContentStyle,
    chatListRightItemContentStyle = props.chatListRightItemContentStyle,
    chatListItemAvatarStyle = props.chatListItemAvatarStyle,
    chatListItemAvatarClassName = props.chatListItemAvatarClassName,
    chatListItemContentClassName = props.chatListItemContentClassName,
    chatListItemTitleClassName = props.chatListItemTitleClassName,
    _props$userMeta = props.userMeta,
    userMeta = _props$userMeta === void 0 ? {
      avatar: DEFAULT_USER_AVATAR
    } : _props$userMeta,
    _props$assistantMeta = props.assistantMeta,
    assistantMeta = _props$assistantMeta === void 0 ? {
      avatar: DEFAULT_AVATAR
    } : _props$assistantMeta,
    _props$chatList = props.chatList,
    chatList = _props$chatList === void 0 ? [] : _props$chatList,
    style = props.style;
  var _useContext = useContext(ConfigProvider.ConfigContext),
    getPrefixCls = _useContext.getPrefixCls;
  var prefixClass = getPrefixCls('pro-chat-list');
  var _useStyle = useStyle(prefixClass),
    wrapSSR = _useStyle.wrapSSR,
    hashId = _useStyle.hashId;
  if (loading) return wrapSSR( /*#__PURE__*/_jsx("div", {
    className: cx("".concat(prefixClass, " ").concat(prefixClass, "-loading"), className, hashId),
    ref: chatListRef,
    children: /*#__PURE__*/_jsx(SkeletonList, {})
  }));
  return wrapSSR( /*#__PURE__*/_jsx("div", {
    className: cx("".concat(prefixClass), className, hashId),
    style: style,
    ref: chatListRef,
    children: chatList.map(function (item) {
      return /*#__PURE__*/_jsx(ChatItem, {
        "data-id": item.id,
        avatar: item.meta || (item.role === 'user' ? userMeta : assistantMeta),
        style: props.chatListItemStyle,
        originData: item,
        placement: item.role === 'user' ? 'right' : 'left',
        time: item.updateAt || item.createAt,
        chatListItemContentStyle: _objectSpread(_objectSpread({}, chatListItemContentStyle), item.role === 'user' ? chatListRightItemContentStyle : chatListLeftItemContentStyle),
        chatListItemTitleStyle: chatListItemTitleStyle,
        chatItemRenderConfig: chatItemRenderConfig,
        chatListItemAvatarStyle: chatListItemAvatarStyle,
        chatListItemAvatarClassName: chatListItemAvatarClassName,
        chatListItemContentClassName: chatListItemContentClassName,
        chatListItemTitleClassName: chatListItemTitleClassName,
        children: /*#__PURE__*/_jsx(MessageComponent, _objectSpread({}, item))
      }, item.id);
    })
  }));
};
export default ChatList;
