// @ant-design/pro-chat/es/components/ProChatInputArea/ActionBar.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { gLocaleObject } from "../../locale";
import { DeleteFilled } from '@ant-design/icons';
import { Flex, Popconfirm, theme } from 'antd';
import cx from 'classnames';
import { jsx as _jsx } from "react/jsx-runtime";
export var ProChatActionBar = function ProChatActionBar(props) {
  var className = props.className,
    prefixClass = props.prefixClass,
    clearMessage = props.clearMessage,
    actionsRender = props.actionsRender,
    locale = props.locale,
    style = props.style;
  var _theme$useToken = theme.useToken(),
    hashId = _theme$useToken.hashId;
  var defaultDoms = [/*#__PURE__*/_jsx(Popconfirm, {
    title: gLocaleObject(locale).clearModalTitle,
    okButtonProps: {
      danger: true
    },
    okText: gLocaleObject(locale).clearDialogue,
    onConfirm: function onConfirm() {
      clearMessage();
    },
    children: /*#__PURE__*/_jsx(DeleteFilled, {
      className: cx("".concat(prefixClass, "-item"), "".concat(prefixClass, "-clear-icon"), hashId),
      title: gLocaleObject(locale).clearCurrentDialogue,
      style: {
        cursor: 'pointer'
      }
    })
  }, 'clear')];
  var renderDoms = actionsRender ? actionsRender === null || actionsRender === void 0 ? void 0 : actionsRender(defaultDoms, props) : defaultDoms;
  if (!renderDoms) return null;
  return /*#__PURE__*/_jsx(Flex, {
    align: 'center',
    style: _objectSpread({
      flexDirection: 'row-reverse'
    }, style),
    className: cx(className, prefixClass, hashId),
    gap: 8,
    children: renderDoms
  });
};
