// @ant-design/pro-chat/es/components/ProChatInputArea/AutoCompleteTextArea.js
import _regeneratorRuntime from "@babel/runtime/helpers/esm/regeneratorRuntime";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["disabled", "mentionRequest"];
import { Mentions } from 'antd';
import { useState } from 'react';

/**
 * Props for the MentionsTextArea component.
 */
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * A component that provides a textarea with mention functionality.
 * It allows users to mention other users by typing "@" followed by their name.
 *
 * @component
 * @example
 * ```tsx
 * <MentionsTextArea
 *   disabled={false}
 *   mentionRequest={fetchMentions}
 *   className="mentions-textarea"
 *   onChange={handleInputChange}
 *   onPressEnter={handleEnterPress}
 *   style={{ width: '100%' }}
 * />
 * ```
 */
export var MentionsTextArea = function MentionsTextArea(props) {
  var disabled = props.disabled,
    mentionRequest = props.mentionRequest,
    rest = _objectWithoutProperties(props, _excluded);
  var _useState = useState([]),
    _useState2 = _slicedToArray(_useState, 2),
    options = _useState2[0],
    setOptions = _useState2[1];
  return /*#__PURE__*/_jsx(Mentions, _objectSpread(_objectSpread({
    className: props.className,
    disabled: disabled,
    autoSize: true,
    rows: rest.rows || 5
  }, rest), {}, {
    style: _objectSpread({
      backgroundColor: 'transparent',
      flex: 1,
      outline: 'none',
      boxShadow: 'none',
      padding: '0 8px',
      margin: 0,
      border: 'none',
      height: 'auto'
    }, rest.style),
    options: options,
    onSelect: function onSelect(value) {
      var _props$onChange;
      (_props$onChange = props.onChange) === null || _props$onChange === void 0 || _props$onChange.call(props, value === null || value === void 0 ? void 0 : value.value);
      setOptions([]);
    },
    prefix: "/",
    onSearch: ( /*#__PURE__*/function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(value, prefix) {
        var _rest$onSearch;
        var result;
        return _regeneratorRuntime().wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (mentionRequest) {
                _context.next = 2;
                break;
              }
              return _context.abrupt("return", []);
            case 2:
              _context.next = 4;
              return mentionRequest === null || mentionRequest === void 0 ? void 0 : mentionRequest(value);
            case 4:
              result = _context.sent;
              setOptions(result || []);
              rest === null || rest === void 0 || (_rest$onSearch = rest.onSearch) === null || _rest$onSearch === void 0 || _rest$onSearch.call(rest, value, prefix);
            case 7:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function (_x, _x2) {
        return _ref.apply(this, arguments);
      };
    }()),
    onPressEnter: function onPressEnter(e) {
      var _props$onPressEnter;
      if (options.length > 0) return;
      (_props$onPressEnter = props.onPressEnter) === null || _props$onPressEnter === void 0 || _props$onPressEnter.call(props, e);
    }
  }));
};
