// @ant-design/pro-chat/es/components/ProChatInputArea/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _regeneratorRuntime from "@babel/runtime/helpers/esm/regeneratorRuntime";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { useRefFunction } from "../../hooks/useRefFunction";
import { SendOutlined } from '@ant-design/icons';
import { Button, ConfigProvider, Flex } from 'antd';
import cx from 'classnames';
import { useContext, useEffect, useMemo, useRef, useState } from 'react';
import AnimationItem from "../Animation";
import { ProChatActionBar } from "./ActionBar";
import { MentionsTextArea } from "./AutoCompleteTextArea";
import StopLoadingIcon from "./StopLoading";
import { useStyle } from "./style";

/**
 * Props for the ChatInputArea component.
 */
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
/**
 * Represents the ChatInputArea component.
 * @param props - The props for the ChatInputArea component.
 * @returns The rendered ChatInputArea component.
 */
export var ChatInputArea = function ChatInputArea(props) {
  var _ref = props || {},
    className = _ref.className,
    placeholder = _ref.placeholder,
    onSend = _ref.onSend,
    inputAreaRender = _ref.inputAreaRender,
    areaRef = _ref.areaRef,
    typing = _ref.typing,
    areaStyle = _ref.areaStyle,
    sendAreaStyle = _ref.sendAreaStyle,
    inputRender = _ref.inputRender,
    sendButtonRender = _ref.sendButtonRender,
    inputAreaProps = _ref.inputAreaProps,
    sendButtonStyle = _ref.sendButtonStyle,
    clearMessage = _ref.clearMessage,
    stopGenerateMessage = _ref.stopGenerateMessage,
    onMessageSend = _ref.onMessageSend,
    actionStyle = _ref.actionStyle,
    locale = _ref.locale,
    chatRef = _ref.chatRef,
    mentionRequest = _ref.mentionRequest,
    actionsRender = _ref.actionsRender;
  var _useContext = useContext(ConfigProvider.ConfigContext),
    getPrefixCls = _useContext.getPrefixCls;
  var _useState = useState(''),
    _useState2 = _slicedToArray(_useState, 2),
    message = _useState2[0],
    setMessage = _useState2[1];
  var isChineseInput = useRef(false);
  var prefixClass = getPrefixCls('pro-chat-input-area');
  var _useStyle = useStyle(prefixClass),
    wrapSSR = _useStyle.wrapSSR,
    hashId = _useStyle.hashId;
  useEffect(function () {
    if (!chatRef || !chatRef.current) return;
    chatRef.current.setInputAreaValue = function (value) {
      setMessage(value);
    };
  }, [chatRef]);
  var send = useRefFunction( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
    var success;
    return _regeneratorRuntime().wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          if (!(onSend && message)) {
            _context.next = 7;
            break;
          }
          _context.next = 3;
          return onSend(message);
        case 3:
          success = _context.sent;
          if (success) {
            onMessageSend(message);
            setMessage('');
          }
          _context.next = 9;
          break;
        case 7:
          onMessageSend(message);
          setMessage('');
        case 9:
        case "end":
          return _context.stop();
      }
    }, _callee);
  })));

  /**
   * Default props for the auto-complete text area.
   */
  var defaultAutoCompleteTextAreaProps = _objectSpread(_objectSpread({
    placeholder: placeholder,
    mentionRequest: mentionRequest
  }, inputAreaProps), {}, {
    className: cx(inputAreaProps === null || inputAreaProps === void 0 ? void 0 : inputAreaProps.className, "".concat(prefixClass, "-input"), hashId),
    value: message,
    onChange: function onChange(value) {
      setMessage(value);
    },
    onCompositionStart: function onCompositionStart() {
      isChineseInput.current = true;
    },
    onCompositionEnd: function onCompositionEnd() {
      isChineseInput.current = false;
    },
    onPressEnter: function onPressEnter(e) {
      if (!typing && !e.shiftKey && !isChineseInput.current) {
        e.preventDefault();
        send();
      }
    }
  });
  var defaultInput = /*#__PURE__*/_jsx(MentionsTextArea, _objectSpread({}, defaultAutoCompleteTextAreaProps));

  /**
   * Supports custom input rendering.
   */
  var inputDom = inputRender ? inputRender === null || inputRender === void 0 ? void 0 : inputRender(defaultInput, function (message) {
    onMessageSend(message);
  }, defaultAutoCompleteTextAreaProps) : defaultInput;

  /**
   * Returns the default button props based on the loading state.
   * If loading is true, the button will have a text type, with the stopGenerateMessage click handler,
   * and the StopLoadingIcon as the icon.
   * If loading is false, the button will have a text type, the send click handler,
   * and the SendOutlined icon as the icon.
   * @returns The default button props.
   */
  var defaultButtonProps = useMemo(function () {
    return typing ? {
      onClick: function onClick() {
        return stopGenerateMessage();
      },
      icon: /*#__PURE__*/_jsx(StopLoadingIcon, {}),
      style: {
        padding: '4px 8px',
        lineHeight: 1,
        height: 28
      }
    } : {
      disabled: !message,
      onClick: function onClick() {
        return send();
      },
      icon: /*#__PURE__*/_jsx(SendOutlined, {}),
      style: {
        padding: '4px 8px',
        lineHeight: 1,
        height: 28
      }
    };
  }, [typing, message]);
  var defaultButtonDom = /*#__PURE__*/_jsx(Button, _objectSpread(_objectSpread({}, defaultButtonProps), {}, {
    style: sendButtonStyle,
    className: cx("".concat(prefixClass, "-button"), hashId),
    children: typing ? '停止生成' : '发送'
  }));
  var buttonDom = sendButtonRender ? sendButtonRender(defaultButtonDom, defaultButtonProps) : defaultButtonDom;
  var defaultInputArea = wrapSSR( /*#__PURE__*/_jsx(ConfigProvider, {
    theme: {
      token: {
        borderRadius: 4
      }
    },
    children: /*#__PURE__*/_jsxs(Flex, {
      ref: areaRef,
      gap: 2,
      vertical: true,
      align: 'right',
      style: areaStyle,
      className: cx("".concat(prefixClass), className, hashId),
      children: [/*#__PURE__*/_jsx(ProChatActionBar, {
        clearMessage: clearMessage,
        actionsRender: actionsRender,
        className: cx(hashId),
        locale: locale,
        prefixClass: "".concat(prefixClass, "-action-bar"),
        style: actionStyle
      }), inputDom, buttonDom ? /*#__PURE__*/_jsx(AnimationItem, {
        animation: true,
        style: sendAreaStyle,
        className: cx("".concat(prefixClass, "-send-area"), hashId),
        children: buttonDom
      }) : null]
    })
  }));
  if (inputAreaRender) {
    return inputAreaRender(defaultInputArea, function (message) {
      onMessageSend(message);
    }, clearMessage);
  }
  return defaultInputArea;
};
export default ChatInputArea;
