// @ant-design/pro-chat/es/components/ProChatInputArea/style.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { resetComponent, useProChatStyleRegister } from "../../utils/useStyle";
var genStyle = function genStyle(token) {
  return _defineProperty(_defineProperty({}, token.componentCls, {
    minHeight: 44,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-end',
    position: 'relative',
    borderTop: "1px solid ".concat(token.colorBorder),
    '&-send-area': {
      display: 'flex',
      justifyContent: 'flex-end',
      width: 'max-content',
      padding: '8px',
      position: 'absolute',
      right: 0,
      bottom: 0
    }
  }), "".concat(token.componentCls, "-action-bar"), _defineProperty({
    position: 'relative',
    display: 'flex',
    padding: "0 ".concat(token.paddingSM, "px")
  }, "& > ".concat(token.componentCls, "-action-bar-item"), {
    padding: token.paddingXS,
    borderRadius: token.borderRadius,
    '&:hover': {
      backgroundColor: token.colorBgTextHover
    }
  }));
};

/**
 * ProchatItem
 * @param prefixCls
 * @returns
 */
export function useStyle(prefixCls) {
  return useProChatStyleRegister('InputArea', function (token) {
    var proChatToken = _objectSpread(_objectSpread({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genStyle(proChatToken), resetComponent(proChatToken)];
  });
}
