// @ant-design/pro-chat/es/components/CopyButton/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["content", "className", "placement"];
import copy from 'copy-to-clipboard';
import { memo } from 'react';
import { useCopied } from "../../hooks/useCopied";
import { CopyOutlined } from '@ant-design/icons';
import { Tooltip } from 'antd';
import { jsx as _jsx } from "react/jsx-runtime";
export var CopyButton = /*#__PURE__*/memo(function (_ref) {
  var content = _ref.content,
    className = _ref.className,
    _ref$placement = _ref.placement,
    placement = _ref$placement === void 0 ? 'right' : _ref$placement,
    props = _objectWithoutProperties(_ref, _excluded);
  var _useCopied = useCopied(),
    copied = _useCopied.copied,
    setCopied = _useCopied.setCopied;
  return /*#__PURE__*/_jsx(Tooltip, _objectSpread(_objectSpread({}, props), {}, {
    title: copied ? '✅ Success' : 'Copy',
    placement: placement,
    children: /*#__PURE__*/_jsx(CopyOutlined, {
      className: className,
      onClick: function onClick() {
        copy(content);
        setCopied();
      }
    })
  }));
});
