// @ant-design/pro-chat/es/components/ProChatAvatar/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["className", "avatar", "title", "size", "shape", "onClick", "prefixCls", "style"];
import { Avatar, theme } from 'antd';
import cx from 'classnames';
import { isEmoji } from "./isEmoji";
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * @module Avatar
 * @description 头像组件，用于显示用户头像信息
 * @exports Avatar
 * @component
 *
 * @param {string} [avatar] - 头像图片的URL或base64数据
 * @param {string} [background] - 头像的背景颜色
 * @param {'circle' | 'square'} [shape='circle'] - 头像的形状，默认为圆形
 * @param {number} [size=40] - 头像的尺寸（像素）
 * @param {string} [title] - 如果未提供头像，则显示的标题文本
 * @param {string} [className] - 自定义的CSS类名
 * @param {Function} [onClick] - 点击头像时触发的回调函数
 * @param {Object} [style] - 自定义的行内样式
 * @param {Object} [props] - 其他传递给AntAvatar组件的属性
 *
 */
export var ProChatAvatar = function ProChatAvatar(_ref) {
  var className = _ref.className,
    avatar = _ref.avatar,
    title = _ref.title,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 40 : _ref$size,
    _ref$shape = _ref.shape,
    shape = _ref$shape === void 0 ? 'circle' : _ref$shape,
    onClick = _ref.onClick,
    prefixCls = _ref.prefixCls,
    style = _ref.style,
    props = _objectWithoutProperties(_ref, _excluded);
  var _theme$useToken = theme.useToken(),
    hashId = _theme$useToken.hashId;
  var isImage = Boolean(avatar && ['/', 'http', 'data:'].some(function (index) {
    return avatar.startsWith(index);
  }));
  var isBase64 = Boolean(avatar === null || avatar === void 0 ? void 0 : avatar.startsWith('data'));
  if (typeof avatar === 'string' && isEmoji(String(avatar))) {
    return /*#__PURE__*/_jsx("div", {
      className: cx("".concat(prefixCls, "-emoji"), hashId),
      children: avatar
    });
  }
  var text = String(isImage ? title : avatar);
  var avatarProps = {
    className: cx(className, "".concat(prefixCls), hashId),
    shape: shape,
    size: size,
    style: onClick ? style : _objectSpread({
      cursor: 'default'
    }, style)
  };
  return isImage ? /*#__PURE__*/_jsx(Avatar, _objectSpread(_objectSpread({
    src: isBase64 ? avatar : /*#__PURE__*/_jsx("img", {
      src: avatar,
      alt: "avatar"
    })
  }, avatarProps), props)) : /*#__PURE__*/_jsx(Avatar, _objectSpread(_objectSpread(_objectSpread({}, avatarProps), props), {}, {
    children: text === null || text === void 0 ? void 0 : text.toUpperCase().slice(0, 2)
  }));
};
