// @ant-design/pro-chat/es/components/ProChat/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { useEffect, useImperativeHandle, useMemo } from 'react';
import RcResizeObserver from 'rc-resize-observer';
import { useRef, useState } from 'react';
import { useRefFunction } from "../../hooks/useRefFunction";
import { BackToBottomButton } from "../..";
import { Flex } from 'antd';
import cx from 'classnames';
import { useChatList } from "../../hooks/useChatList";
import ChatList from "../ChatList";
import ChatInputArea from "../ProChatInputArea";

/**
 * Represents an instance of the ProChat component.
 * @template T - The type of additional data associated with each chat message.
 */

/**
 * ProChatProps 是 ProChat 组件的属性类型定义。
 * @template T - 聊天记录的数据类型
 */
/**
 * Represents the props for the ProChat component.
 *
 * @template T - The type of the chat message.
 * @template Params - The type of the chat parameters.
 */
import { Fragment as _Fragment } from "react/jsx-runtime";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
/**
 * 对话组件的属性接口
 */

export function ProChat(props) {
  var _areaHtml$current2;
  var style = props.style,
    className = props.className,
    chatItemRenderConfig = props.chatItemRenderConfig,
    backToBottomConfig = props.backToBottomConfig,
    inputRender = props.inputRender,
    inputAreaRender = props.inputAreaRender,
    inputAreaProps = props.inputAreaProps,
    chatRef = props.chatRef,
    userProfile = props.userProfile,
    sendButtonRender = props.sendButtonRender,
    placeholder = props.placeholder,
    styles = props.styles,
    request = props.request,
    onChatsChange = props.onChatsChange,
    userMeta = props.userMeta,
    assistantMeta = props.assistantMeta,
    classNames = props.classNames,
    _sendMessageRequest = props.sendMessageRequest;
  var chatListContainerRef = useRef(null);
  var areaHtml = useRef(null);
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    isInitRender = _useState2[0],
    setIsRender = _useState2[1];
  var _useState3 = useState('100%'),
    _useState4 = _slicedToArray(_useState3, 2),
    height = _useState4[0],
    setHeight = _useState4[1];
  useEffect(function () {
    setIsRender(true);
  }, []);
  var _useChatList = useChatList({
      chatList: props.chatList,
      loading: props.loading,
      initialChatList: props.initialChatsList,
      onChatsChange: onChatsChange ? onChatsChange : undefined,
      helloMessage: /*#__PURE__*/_jsx(_Fragment, {
        children: props.helloMessage || '让我们开始对话吧'
      }),
      userProfile: userProfile,
      request: request ? function () {
        return request(_objectSpread(_objectSpread({}, props.config), props.params));
      } : undefined,
      transformToChatMessage: props.transformToChatMessage,
      sendMessageRequest: function sendMessageRequest(message) {
        return _sendMessageRequest ? _sendMessageRequest(message, _objectSpread(_objectSpread({}, props.config), props.params)) : undefined;
      }
    }),
    chatList = _useChatList.chatList,
    loading = _useChatList.loading,
    isLoadingMessage = _useChatList.isLoadingMessage,
    setMessageItem = _useChatList.setMessageItem,
    stopGenerateMessage = _useChatList.stopGenerateMessage,
    clearMessage = _useChatList.clearMessage,
    sendMessage = _useChatList.sendMessage,
    genMessageRecord = _useChatList.genMessageRecord;
  var getChatList = useRefFunction(function () {
    return chatList;
  });
  useImperativeHandle(chatRef, function () {
    return {
      stopGenerateMessage: stopGenerateMessage,
      clearMessage: clearMessage,
      sendMessage: sendMessage,
      getChatList: getChatList,
      setMessageItem: setMessageItem,
      genMessageRecord: genMessageRecord,
      scrollToBottom: function scrollToBottom() {
        var _current, _chatListContainerRef;
        chatListContainerRef === null || chatListContainerRef === void 0 || (_current = chatListContainerRef.current) === null || _current === void 0 || _current.scrollTo({
          behavior: 'smooth',
          left: 0,
          top: ((_chatListContainerRef = chatListContainerRef.current) === null || _chatListContainerRef === void 0 ? void 0 : _chatListContainerRef.scrollHeight) || 99999
        });
      }
    };
  }, [chatRef]);
  var scrollToBottom = useMemo(function () {
    if (chatListContainerRef.current) {
      return function () {
        chatListContainerRef.current.scrollTo({
          behavior: 'smooth',
          left: 0,
          top: chatListContainerRef.current.scrollHeight || 99999
        });
      };
    }
    return function () {};
  }, [chatListContainerRef]);
  useEffect(function () {
    if (chatListContainerRef.current) {
      chatListContainerRef.current.scrollTo({
        behavior: 'smooth',
        left: 0,
        top: chatListContainerRef.current.scrollHeight
      });
    }
  }, [chatList]);
  var backBottomDom = useMemo(function () {
    var _areaHtml$current;
    if (!isInitRender) return null;
    return /*#__PURE__*/_jsx(BackToBottomButton, _objectSpread(_objectSpread({
      offsetTop: ((_areaHtml$current = areaHtml.current) === null || _areaHtml$current === void 0 ? void 0 : _areaHtml$current.clientHeight) || 0,
      target: function target() {
        return chatListContainerRef.current;
      }
    }, backToBottomConfig), {}, {
      onClick: scrollToBottom
    }));
  }, [isInitRender]);
  return /*#__PURE__*/_jsx(RcResizeObserver, {
    onResize: function onResize(e) {
      if (e.height !== height) {
        setHeight(e.height - 1);
      }
    },
    children: /*#__PURE__*/_jsxs(Flex, {
      className: cx(className),
      style: _objectSpread({
        maxHeight: '100vh',
        height: '100%',
        position: 'relative'
      }, style),
      vertical: true,
      children: [props.beforeChatListDom, /*#__PURE__*/_jsx(ChatList, {
        chatListRef: chatListContainerRef,
        chatList: chatList,
        userMeta: userMeta,
        assistantMeta: assistantMeta,
        loading: loading,
        chatItemRenderConfig: chatItemRenderConfig,
        style: _objectSpread(_objectSpread({}, styles === null || styles === void 0 ? void 0 : styles.chatList), {}, {
          height: height - (((_areaHtml$current2 = areaHtml.current) === null || _areaHtml$current2 === void 0 ? void 0 : _areaHtml$current2.clientHeight) || 0) || '100%'
        }),
        className: classNames === null || classNames === void 0 ? void 0 : classNames.chatList,
        chatListItemClassName: classNames === null || classNames === void 0 ? void 0 : classNames.chatListItem,
        chatListItemContentClassName: classNames === null || classNames === void 0 ? void 0 : classNames.chatListItemContent,
        chatListItemTitleClassName: classNames === null || classNames === void 0 ? void 0 : classNames.chatListItemTitle,
        chatListItemAvatarClassName: classNames === null || classNames === void 0 ? void 0 : classNames.chatListItemAvatar
        // styles
        ,
        chatListItemStyle: styles === null || styles === void 0 ? void 0 : styles.chatListItem,
        chatListLeftItemContentStyle: styles === null || styles === void 0 ? void 0 : styles.chatListLeftItemContent,
        chatListRightItemContentStyle: styles === null || styles === void 0 ? void 0 : styles.chatListRightItemContent,
        chatListItemContentStyle: styles === null || styles === void 0 ? void 0 : styles.chatListItemContent,
        chatListItemTitleStyle: styles === null || styles === void 0 ? void 0 : styles.chatListItemTitle,
        chatListItemExtraStyle: styles === null || styles === void 0 ? void 0 : styles.chatListItemExtra,
        chatListItemAvatarStyle: styles === null || styles === void 0 ? void 0 : styles.chatListItemAvatar
      }), backBottomDom, /*#__PURE__*/_jsx(ChatInputArea, {
        className: classNames === null || classNames === void 0 ? void 0 : classNames.chatInputArea,
        typing: isLoadingMessage,
        placeholder: placeholder || '请输入消息...',
        onMessageSend: sendMessage,
        mentionRequest: props.autocompleteRequest,
        stopGenerateMessage: stopGenerateMessage,
        clearMessage: clearMessage,
        areaRef: areaHtml,
        sendButtonStyle: styles === null || styles === void 0 ? void 0 : styles.chatSendButton,
        actionsRender: props.actionsRender,
        sendButtonRender: sendButtonRender,
        inputAreaRender: inputAreaRender,
        inputRender: inputRender,
        inputAreaProps: inputAreaProps,
        chatRef: chatRef,
        actionStyle: styles === null || styles === void 0 ? void 0 : styles.chatInputAction,
        sendAreaStyle: styles === null || styles === void 0 ? void 0 : styles.chatSendAreaStyle,
        areaStyle: styles === null || styles === void 0 ? void 0 : styles.chatInputArea
      })]
    })
  });
}
